### ISSUE TITLE
How to use component properties/methods when paypal callbacks (payment/onAuthorize) ?

### ISSUE THREAD

**Issue by amaneer94** [paypal/paypal-checkout#480]

**Comment by amaneer94** [comment]
If we want to use other properties or call methods of component. Component's 'this' keyword is not working inside definition of both callback methods (payment/onAuthorize).

**Comment by bluepnume** [comment]
What specifically are you trying to use?

**Comment by amaneer94** [comment]
I have placed <paypal-button> in an angular component. When onAuthorize() is triggered, i want to emit an angular EventEmitter ...

**Comment by amaneer94** [comment]
`public onAuthorize(data, actions) {
	return actions.payment.execute().then(function() {
	this.mPaymentSuccessEmitter.emit();
	});
}`

It says that .. cannot read 'emit' of undefined.

**Comment by amaneer94** [comment]
I think context 'this' is changed in `onAuthorize(data, actions) ....` But don't know how to access component's properties.

**Comment by bluepnume** [comment]
OK. You have a few choices:

1. Use an arrow function:

```javascript
public onAuthorize(data, actions) {
    return actions.payment.execute().then(() => {
        this.mPaymentSuccessEmitter.emit();
    });
}
```

2. Alias `self = this`:

```javascript
public onAuthorize(data, actions) {
    var self = this;

    return actions.payment.execute().then(function() {
        self.mPaymentSuccessEmitter.emit();
    });
}
```

Going to close this issue though, as this is a [general javascript problem](https://stackoverflow.com/questions/20279484/how-to-access-the-correct-this-inside-a-callback) rather than something specific to checkout.js

**Comment by amaneer94** [comment]
I understand this problem, but when I tried to `console.log(this);` inside It shows me 'props' object rather than that Component.

**Comment by amaneer94** [comment]
As i am following this.

https://github.com/paypal/paypal-checkout/issues/368

 (xcomponent concept).

(without AOT)

**Comment by bluepnume** [comment]
Can you paste the entire class?

**Comment by amaneer94** [comment]
`import { CartModel } from '../../models';
import { CartOrderManagementService } from './../../services';

@Component({
	selector: 'paypal-checkout-button',
	templateUrl: 'paypal-checkout-button.component.html',
	styleUrls: ['./paypal-checkout-button.component.css']
})

export class PaypalCheckoutButtonComponent implements OnInit {
	@Input() cart: CartModel;//if cart is not provided .. if will fetch from cart service ...
	paymentMethods = constants.paymentMethods;
	constructor(
		private _cartService: CartOrderManagementService,
	) {

	}
	ngOnInit() {
		if(!this.cart) {
			this.cart = this._cartService.getCart();
		}
		this.cart.paymentMethod = "paypal";

	}
	////////paypal////////////////////
	@Input() env = 'sandbox'
	@Input() client = {
		sandbox:    'AW6YL6v3VfNzgPTfS_pjymBFezV7JUiBtZKGMl6J2oRFDEILkgi6Bn1R9e34Bo58FWE8A-e_8nlx5lcZ',
		production: 'xxxx'
	}
	@Input() commit = true
	private payment(data, actions) {
		return actions.payment.create({
			transactions: [
				{
					amount: { total: ''+this.cart.grandTotal, currency: 'USD' }
				}
			]
		});
	}
	private onAuthorize(data, actions) {
		console.log("Authorized: ", this, data, actions);
		return actions.payment.execute().then(function() {
			console.log("console log this !", this);
		});
	}	
}
`

**Comment by amaneer94** [comment]
this is component html template

`<paypal-button fxFlex [props]="{
    env: env,
    client: client,
    commit: commit,
    payment: payment,
    onAuthorize: onAuthorize
}"></paypal-button>
`

**Comment by amaneer94** [comment]
`import { NgModule } from '@angular/core';
import { NgModule, Component, ElementRef, NgZone } from '@angular/core';
import { SharedModule } from '../../shared/shared-module/shared.module';
import { routedComponents, FrontendRoutingModule } from './frontend-routing.module';

import * as ngCore from '@angular/core';
declare let paypal: any;
const PayPalButtonModule = paypal.Button.driver('angular2', { Component, NgModule, ElementRef, NgZone });
@NgModule({
    imports: [
        FrontendRoutingModule,
        SharedModule
        SharedModule,
        PayPalButtonModule
    ],
............`

**Comment by amaneer94** [comment]
and checkout.js (cdn) is included in index.html

**Comment by bluepnume** [comment]
Can you try something along the lines of:

```
import { CartModel } from '../../models';
import { CartOrderManagementService } from './../../services';

@Component({
	selector: 'paypal-checkout-button',
	templateUrl: 'paypal-checkout-button.component.html',
	styleUrls: ['./paypal-checkout-button.component.css']
})

export class PaypalCheckoutButtonComponent implements OnInit {
	@Input() cart: CartModel;//if cart is not provided .. if will fetch from cart service ...
	paymentMethods = constants.paymentMethods;
	constructor(
		private _cartService: CartOrderManagementService,
	) {
            this.onAuthorize = this.onAuthorize.bind(this);
	}
	ngOnInit() {
		if(!this.cart) {
			this.cart = this._cartService.getCart();
		}
		this.cart.paymentMethod = "paypal";

	}
	////////paypal////////////////////
	@Input() env = 'sandbox'
	@Input() client = {
		sandbox:    'AW6YL6v3VfNzgPTfS_pjymBFezV7JUiBtZKGMl6J2oRFDEILkgi6Bn1R9e34Bo58FWE8A-e_8nlx5lcZ',
		production: 'xxxx'
	}
	@Input() commit = true
	private payment(data, actions) {
		return actions.payment.create({
			transactions: [
				{
					amount: { total: ''+this.cart.grandTotal, currency: 'USD' }
				}
			]
		});
	}
	private onAuthorize(data, actions) {
		console.log("Authorized: ", this, data, actions);
		return actions.payment.execute().then(() => {
			console.log("console log this !", this);
		});
	}	
}
```

**Comment by usamamalik75** [comment]
@bluepnume 
Its working Thanks
but one problem after payment successful Router object undefined
some sample code

onAuthorize(data: any, actions: any) {
debugger
return actions.payment.execute().then(function () {
// Show a success page to the buyer
console.log("onAuthorize");
//// after successful payment "router" object show undefined
this.router.navigate(['abc',id]);
});

}

**Comment by bluepnume** [comment]
You have the exact same problem -- you can't use `this` in a regular callback function. You need to use `return actions.payment.execute().then(() => {` or alias `var self = this`

**Comment by amaneer94** [comment]
I also tried `var self = this;`,   but here `'this'` is used inside `onAuthorize(data: any, actions: any) ...` it threw props object in `this`, NOT angular component's `this`.

**Comment by amaneer94** [comment]
But amazingly `this` worked for me when I used AOT solution for paypal-express-checkout for angular 2+

**Comment by usamamalik75** [comment]
@amaneer94 
i clone your project but i don't see any code which access angular object after success paypal payment

your code is 

>  onAuthorize: (data, actions) => {
>       // show success page
>     }

**Comment by usamamalik75** [comment]
@bluepnume 
i follow your suggestion but issue still
![image](https://user-images.githubusercontent.com/33087810/32337301-6c7ae62a-c013-11e7-972c-0ae54eb00bf1.png)
please see in image this again undefine 
please solve my problem how i navigate to other componet(URL)
Thanks

**Comment by bluepnume** [comment]
You have to do `var self = this` *outside* of any callbacks, at a point where `this` has the expected value.

**Comment by usamamalik75** [comment]
@bluepnume 
i want redirect to other URl(component) when payment successfull
please suggest any other idea how i achieve this

**Comment by usamamalik75** [comment]
@bluepnume  @amaneer94 
At last i got solution
```
public onAuthorize=(data: any, actions: any)=>{
        debugger
        this.router.navigate(['abc',id])
    } 
```

**Comment by amaneer94** [comment]
@usamamalik75 I referred you to AOT solution for paypal, if you try 
>  onAuthorize: (data, actions) => {
>       // show success page
>       console.log(this);
>     }
you will see this time it would be angular component, and you will be free to access any property of component directly or using var self = this; technique.

**Comment by usamamalik75** [comment]
@amaneer94  
Thanks to prefer AOT solution both result are same :)
But problem now 
my account have 13 USD
but i set my shopping amount 15 USD
in this cause paypal response show error in console 
![image](https://user-images.githubusercontent.com/33087810/32407044-9202aa26-c1a4-11e7-8c10-2721ec346f98.png)

Authorize method not work what method i can use to handle when payment not successfull?
payment
authorize and third one is error?

**Comment by ThakkarKhushbu** [comment]
I have follow below for integrate paypal to my angular 5 project

https://developer.paypal.com/demo/checkout/#/pattern/client

Have same issue above , on 
return actions.payment.execute().then(function() {
                  // want to access function 
                    window.alert('Payment Complete!');
                });

not able to access this here . I want to make one api call based on payment success.


Tried both solution

public onAuthorize(data, actions) {
    return actions.payment.execute().then(() => {
        this.mPaymentSuccessEmitter.emit();
    });
}
Alias self = this:
public onAuthorize(data, actions) {
    var self = this;

    return actions.payment.execute().then(function() {
        self.mPaymentSuccessEmitter.emit();
    });
}

still getting error ,  
![paypal](https://user-images.githubusercontent.com/30835218/34822426-ea4fc79a-f6ec-11e7-88d3-e14cd4994d22.png)


{error: "Error: TypeError: _this3.props.onError is not a fu…://www.paypalobjects.com/api/checkout.js:3080:13)", timestamp: 1515668444833, windowID: "2c885cbc12", pageID: "9a43f9dfda", prev_corr_ids: "8d795cbc1430f, 8d795cbc1430f", …}
